pdf2htmlEX.1
.TH pdf2htmlEX 1 "Aug 31, 2012" "pdf2htmlEX 0.1"
.SH NAME
.PP
.nf
  pdf2htmlEX \- converts PDF to HTML without losing text and format.
.fi

.SH USAGE
.PP
.nf
  pdf2htmlEX [options] <input\-filename> [<output\-filename>]
.fi

.SH DESCRIPTION
.PP
pdf2htmlEX is a utility that converts PDF files to HTML files.

pdf2htmlEX tries its best to render the PDF precisely, maintain proper styling, while retaining text and optmizing for Web.

Fonts are extracted form PDF and then embedded into HTML (Type 3 fonts are not supported). Text in the converted HTML file is usually selectable and copyable. 

Other objects are rendered as images and also embedded.

.SH OPTIONS
.TP
.B --help
Show all options
.TP
.B -v, --version
Show copyright and version
.TP
.B -o, --owner-password <password>
Specify owner password
.TP
.B -u, --user-password <password>
Specify user password
.TP
.B --dest-dir <dir> (Default: ".")
Specify destination folder
.TP
.B -f, --first-page <num> (Default: 1)
Specify the first page to process
.TP
.B -l, --last-page <num> (Default: last page)
Specify the last page to process
.TP
.B --zoom <ratio> (Default: 1.0)
Specify the zoom ratio of the HTML file
.TP
.B --hpdi <dpi>, --vpdi <dpi> (Default: 144)
Specify the horizontal and vertical DPI for images
.TP
.B --process-nontext <0|1> (Default: 1)
Whether to process non-text objects (as images)
.TP
.B --single-html <0|1> (Default: 1)
Whether to embed everything into one HTML file.

If switched out, there will be several files generated along with the HTML file including files for fonts, css, images.
.TP
.B --split-pages <0|1> (Default: 0)
If turned on, each page is saved in a separated files, also the generated css file will be store separatedly as if single-html=0

The output files will be named as <output-filename>0.page, <output-filename>1.page, ...
.TP
.B --embed-base-font <0|1> (Default: 1)
Whether to embed base 14 fonts.

There are several base font defined in PDF standards, which are supposed to be provided by the PDF reader.

If this switch is on, local matched font will be used and embedded; otherwise only font names are exported such that web browsers may try to find proper fonts themselves.
.TP
.B --embed-external-font <0|1> (Default: 0)
Similar as above but for non-base fonts.
.TP
.B --decompose-ligature <0|1> (Default: 0)
Decompose ligatures. For example 'fi' -> 'f''i'.
.TP
.B --heps <len>, --veps <len> (Default: 1)
Specify the maximum tolerable horizontal/vertical offset (in pixels).

pdf2htmlEX would try to optimize the generated HTML file moving Text within this distance.
.TP
.B --space-threshold <ratio> (Default: 1.0/6)
pdf2htmlEX would insert a whitespace character ' ' if the distance between two consecutive letters in the same line is wider than ratio * font_size.
.TP
.B --font-size-multiplier <ratio> (Default: 10)
Many web browsers limit the minimum font size, and many would round the given font size, which results in incorrect rendering.

Specify a ratio greater than 1 would resolve this issue.

For some versions of Firefox, however, there will be a problem when the font size is too large, in which case a smaller value should be specified here.
.TP
.B --tounicode <-1|0|1> (Default: 0)
A ToUnicode map may be provided for each font in PDF which indicates the 'meaning' of the characters. However often there is better "ToUnicode" info in Type 0/1 fonts, and sometimes the ToUnicode map provided is wrong. 

If this value is set to 1, the ToUnicode Map is always applied, if provided in PDF, and characters may not render correctly in HTML if there are collisions.

If set to -1, a customized map is used such that rendering will be correct in HTML (visually the same), but you may not get correct characters by select & copy & paste.

If set to 0, pdf2htmlEX would try it best to balance the two methods above.
.TP
.B --space-as-offset <0|1> (Default: 0)
Treat space characters as offsets, which may increase the size of the output.

Turn it on if space characters are not displayed correctly, or you want to remove positional spaces.
.TP
.B --css-filename <filename> (Default: "")
Specify the filename of the generated css file, if not embedded.

If it's empty, the file name will be determined automatically.
.TP
.B --font-suffix <suffix> (Default: ".ttf"), --font-format <format> (Default: "truetype")
Specify the suffix and format of fonts extracted from the PDF file. They should be consistent.
.TP
.B --debug <0|1> (Default: 0)
Show debug information.
.TP
.B --clean-tmp <0|1> (Default: 1)
If switched off, intermediate files won't be cleaned in the end.

.SH EXAMPLE
.TP
.B pdf2htmlEX /path/to/file.pdf
Convert file.pdf into file.html
.TP
.B pdf2htmlEX --tmp-dir tmp --clean-tmp 0 --debug 1 /path/to/file.pdf
Convert file.pdf and leave all intermediate files.
.TP
.B pdf2htmlEX --dest-dir out --single-html 0 --debug 1 /path/to/file.pdf
Convert file.pdf into out/file.html and leave font/image files separated.

.SH COPYRIGHT
.PP
Copyright 2012 Lu Wang <coolwanglu@gmail.com>

pdf2htmlEX is GPLv2 & GPLv3 dual licensed

.SH AUTHOR
.PP
pdf2htmlEX is written by Lu Wang <coolwanglu@gmail.com>

.SH SEE ALSO
.TP
Home page
http://github.com/coolwanglu/pdf2htmlEX


ChangeLog
Latest v0.3

* New options: --data-dir --split-pages
* Manifest for customized HTML generation
* Smooth scrolling
* Completely removed dependency of boost
* Removed dependency of boost::filesystem
* Relaxed dependency on c++11. Now can be built with GCC 4.4.6
* Removed dependency of boost::format and boost::algorithm
* New option --space-as-offset
* A font preprocessor, for solving encoding problems
* Better HTML optimization, states are reused
* HTML should work when Javascript is disabled

v0.2
2012.09.06

* Fontforge is now linked with, not called with scripts
* Better accuracy of rendering, with a new line model
* New option --decompose-ligature

v0.1 
2012.08.28

* The first release 


src/pdf2htmlEX.cc
// pdftohtmlEX.cc
//
// Copyright (C) 2012 Lu Wang coolwanglu<at>gmail.com

#include <cstdio>
#include <cstdlib>
#include <cstddef>
#include <cstring>
#include <ctime>
#include <string>
#include <limits>
#include <iostream>
#include <getopt.h>

#include <goo/GooString.h>

#include <Object.h>
#include <PDFDoc.h>
#include <PDFDocFactory.h>
#include <GlobalParams.h>

#include "HTMLRenderer.h"
#include "Param.h"
#include "pdf2htmlEX-config.h"
#include "ArgParser.h"

using namespace std;
using namespace pdf2htmlEX;

Param param;
ArgParser argparser;

void show_usage_and_exit(const char * dummy = nullptr)
{
    cerr << "pdftohtmlEX version " << PDF2HTMLEX_VERSION << endl;
    cerr << endl;
    cerr << "Copyright 2012 Lu Wang (coolwanglu<at>gmail.com)" << endl;
    cerr << endl;
    cerr << "Usage: pdf2htmlEX [Options] <input.pdf> [<output.html>]" << endl;
    cerr << endl;
    cerr << "Options:" << endl;
    argparser.show_usage(cerr);
    cerr << endl;
    exit(EXIT_FAILURE);
}

void parse_options (int argc, char **argv)
{
    argparser
        .add("help,h", "show all options", &show_usage_and_exit)
        .add("version,v", "show copyright and version info", &show_usage_and_exit)

        .add("owner-password,o", &param.owner_password, "", "owner password (for encrypted files)")
        .add("user-password,u", &param.user_password, "", "user password (for encrypted files)")

        .add("dest-dir", &param.dest_dir, ".", "specify destination directory")
        .add("data-dir", &param.data_dir, PDF2HTMLEX_DATA_PATH, "specify data directory")

        .add("first-page,f", &param.first_page, 1, "first page to process")
        .add("last-page,l", &param.last_page, numeric_limits<int>::max(), "last page to process")

        .add("zoom", &param.zoom, 1.0, "zoom ratio")
        .add("hdpi", &param.h_dpi, 144.0, "horizontal DPI for non-text")
        .add("vdpi", &param.v_dpi, 144.0, "vertical DPI for non-text")

        .add("process-nontext", &param.process_nontext, 1, "process nontext objects")
        .add("single-html", &param.single_html, 1, "combine everything into one single HTML file")
        .add("split-pages", &param.split_pages, 0, "split pages into separated files")
        .add("embed-base-font", &param.embed_base_font, 0, "embed local matched font for base 14 fonts in the PDF file")
        .add("embed-external-font", &param.embed_external_font, 0, "embed local matched font for external fonts in the PDF file")
        .add("decompose-ligature", &param.decompose_ligature, 0, "decompose ligatures, for example 'fi' -> 'f''i'")

        .add("heps", &param.h_eps, 1.0, "max tolerated horizontal offset (in pixels)")
        .add("veps", &param.v_eps, 1.0, "max tolerated vertical offset (in pixels)")
        .add("space-threshold", &param.space_threshold, (1.0/8), "distance no thiner than (threshold * em) will be considered as a space character")
        .add("font-size-multiplier", &param.font_size_multiplier, 10.0, "setting a value greater than 1 would increase the rendering accuracy")
        .add("tounicode", &param.tounicode, 0, "Specify how to deal with ToUnicode map, 0 for auto, 1 for forced, -1 for disabled")
        .add("space-as-offset", &param.space_as_offset, 0, "treat space characters as offsets")

        .add("css-filename", &param.css_filename, "", "Specify the file name of the generated css file")
        .add("font-suffix", &param.font_suffix, ".ttf", "suffix for extracted font files")
        .add("font-format", &param.font_format, "opentype", "format for extracted font files")

        .add("debug", &param.debug, 0, "output debug information")
        .add("clean-tmp", &param.clean_tmp, 1, "clean temporary files after processing")
        .add("", &param.input_filename, "", "")
        .add("", &param.output_filename, "", "")
        ;

    try
    {
        argparser.parse(argc, argv);
    }
    catch(const char * s)
    {
        // if s == "", getopt_long would have printed the error message
        if(s && s[0])
        {
            cerr << "Error when parsing the arguments:" << endl;
            cerr << s << endl;
        }
        exit(EXIT_FAILURE);
    }
    catch(const std::string & s)
    {
        // if s == "", getopt_long would have printed the error message
        if(s != "")
        {
            cerr << "Error when parsing the arguments:" << endl;
            cerr << s << endl;
        }
        exit(EXIT_FAILURE);
    }
}

int main(int argc, char **argv)
{
    parse_options(argc, argv);
    if (param.input_filename == "")
    {
        cerr << "Missing input filename" << endl;
        exit(EXIT_FAILURE);
    }

    //prepare the directories
    {
        char buf[] = "/tmp/pdf2htmlEX-XXXXXX";
        auto p = mkdtemp(buf);
        if(p == nullptr)
        {
            cerr << "Cannot create temp directory" << endl;
            exit(EXIT_FAILURE);
        }
        param.tmp_dir = buf;
    }

    if(param.debug)
        cerr << "temporary dir: " << (param.tmp_dir) << endl;

    try
    {
        create_directories(param.dest_dir);
    }
    catch (const string & s)
    {
        cerr << s << endl;
        exit(EXIT_FAILURE);
    }

    bool finished = false;
    // read config file
    globalParams = new GlobalParams();
    // open PDF file
    PDFDoc *doc = nullptr;
    try
    {
        {
            GooString * ownerPW = (param.owner_password == "") ? (nullptr) : (new GooString(param.owner_password.c_str()));
            GooString * userPW = (param.user_password == "") ? (nullptr) : (new GooString(param.user_password.c_str()));
            GooString fileName(param.input_filename.c_str());

            doc = PDFDocFactory().createPDFDoc(fileName, ownerPW, userPW);

            delete userPW;
            delete ownerPW;
        }

        if (!doc->isOk()) {
            throw "Cannot read the file";
        }

        // check for copy permission
        if (!doc->okToCopy()) {
            throw "Copying of text from this document is not allowed.";
        }

        param.first_page = min(max(param.first_page, 1), doc->getNumPages());
        param.last_page = min(max(param.last_page, param.first_page), doc->getNumPages());

        if(param.output_filename == "")
        {
            const string s = get_filename(param.input_filename);

            if(get_suffix(param.input_filename) == ".pdf")
            {
                if(param.split_pages)
                    param.output_filename = s.substr(0, s.size() - 4);
                else
                    param.output_filename = s.substr(0, s.size() - 4) + ".html";

            }
            else
            {
                if(param.split_pages)
                    param.output_filename = s;
                else
                    param.output_filename = s + ".html";
                
            }
        }
        if(param.css_filename == "")
        {
            const string s = get_filename(param.input_filename);

            if(get_suffix(param.input_filename) == ".pdf")
            {
                param.css_filename = s.substr(0, s.size() - 4) + ".css";
            }
            else
            {
                if(!param.split_pages)
                    param.css_filename = s + ".css";
            }
        }

        HTMLRenderer * htmlOut = new HTMLRenderer(&param);
        htmlOut->process(doc);
        delete htmlOut;

        finished = true;
    }
    catch (const char * s)
    {
        cerr << "Error: " << s << endl;
    }
    catch (const string & s)
    {
        cerr << "Error: " << s << endl;
    }

    // clean up
    if(doc) delete doc;
    if(globalParams) delete globalParams;

    // check for memory leaks
    Object::memCheck(stderr);
    gMemReport(stderr);

    exit(finished ? (EXIT_SUCCESS) : (EXIT_FAILURE));

    return 0;
}
